include/misc/sem.h

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ *
+ * @brief public sys_sem APIs.
+ */
+
+#ifndef ZEPHYR_INCLUDE_MISC_SEM_H_
+#define ZEPHYR_INCLUDE_MISC_SEM_H_
+
+/*
+ * sys_sem exists in user memory working as counter semaphore for
+ * user mode thread when user mode enabled. When user mode isn't
+ * enabled, sys_sem behaves like k_sem.
+ */
+
+#include <kernel.h>
+#include <atomic.h>
+#include <zephyr/types.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * sys_sem structure
+ */
+struct sys_sem {
+#ifdef CONFIG_USERSPACE
+	struct k_futex futex;
+	int limit;
+#else
+	struct k_sem kernel_sem;
+#endif
+};
+
+/**
+ * @brief Initialize a semaphore.
+ *
+ * This routine initializes a semaphore instance, prior to its first use.
+ *
+ * @param sem Address of the semaphore.
+ * @param initial_count Initial semaphore count.
+ * @param limit Maximum permitted semaphore count.
+ *
+ * @retval 0 Initial success.
+ * @retval -EINVAL Bad parameters, the value of limit should be located in
+ *         (0, INT_MAX] and initial_count shouldn't be greater than limit.
+ */
+int sys_sem_init(struct sys_sem *sem, unsigned int initial_count,
+		unsigned int limit);
```

Comment:
Design note: do we want "limit" here?  Why?  Who uses it?  This is IMHO sort of an API wart.  It provides very little value, and has a runtime cost in both CPU cycles and memory bytes.  If we're going to invent a new semaphore API, my vote is we junk it.

Reply:
@andyross I'm confused here. limit is the value in which further attempts to take result in blocking. if I have a four count semaphore, limit=4 I don't see a way around this.

Reply:
It's not, it's the value at which further attempts to **give** result in blocking.  And that's a werid quirk of our API.  Normally semaphores are documented to handle arbitrary up counts, the standard definition of the give/up/V/release/whatever operation almost never talks about the possibility of it blocking.  Check the docs on linux sem_post(), win32 ReleaseSemaphore(), the wikipedia "Semaphore (Programming)" page, et. al.

I'm sure there are other OSes somewhere with this behavior, we can't be the only one.  But it's just a weird thing.  And we're carrying it around as a 4 byte per semaphore cost and a few cycles on every op, and I think it's needless.

Per another comment here, I'd prefer to see us wrap **this** API as the basis for k_sem, and put the weird legacy thing there.  But that requires removing it here.

Reply:
Actually, I just checked: our behavior is even weirder.  We don't, in fact, block if we hit the limit.  We just throw the up value away.  That's actually IMHO worse, as it makes the number of calls to give and take mismatched when a limit is hit.  The "writing" half of an app will expect to have woken up N threads to handle N events or whatever, but in fact the number of running threads is less due to the limit.

Reply:
>> limit is the value in which further attempts to take result in blocking. if I have a four count semaphore, limit=4 I don't see a way around this.

@andyross  I mis-read the code. Taking decrements the count. Giving increases it. If you give more than the limit, the count stays at the limit.

This all seems working as intended to me.

If we want to change the semantics of our semaphore APIs, this is not the PR to do it.

Reply:
It's literally a new semaphore, though.  We can have any API we want.  Why pick the old weird one except pure inertia?  Like I said, we can put the limit into the k_sem wrapper.

Reply:
> We can have any API we want. Why pick the old weird one except pure inertia? 

I wanted them to have exactly the same semantics, so it isn't difficult to merge them into a single API later.

Reply:
And I want limit to not be carried forward as cruft, **because** it makes it harder to merge them to a cruft-free API later. :)

But again, this isn't a -1.  I'm just saying that putting this API in here means one extra place we need to deprecate later.

Reply:
> But again, this isn't a -1. I'm just saying that putting this API in here means one extra place we need to deprecate later.

I at least am all right with that.
k_sem has been like this for at least two years, ever since Ben did the kernel API overhaul. I'm actually kinda shocked you only just noticed this now. I don't think this PR is the place to redesign semaphores, not this late in the dev cycle for 2.0. I also think it's more trouble than it's worth to remove 'limit' from k_sem.